src/raft/raft.go

```diff
@@ -41,25 +46,319 @@ type ApplyMsg struct {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm--
+	return
+}
+
+func (rf *Raft) setTerm(term int) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm = term
+	return
+}
+
+func (rf *Raft) getMajority() int {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	return rf.majority
+
+}
+
+func (rf *Raft) setLastContact() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.lastContact = time.Now()
+}
+
+func (rf *Raft) LastContact() time.Time {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	contact := rf.lastContact
+	return contact
+}
+
+func (rf *Raft) getID() (ID int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	ID = rf.me
+	return
+}
+
+func (rf *Raft) getTerm() (term int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	term = rf.currentTerm
+	return
+}
+
 // return currentTerm and whether this server
 // believes it is the leader.
 func (rf *Raft) GetState() (int, bool) {
-
-	var term int
-	var isleader bool
+	term := rf.getTerm()
+	isLeader := rf.isLeader()
 	// Your code here.
-	return term, isleader
+	return term, isLeader
+}
+
+// This is our main loop. It checks the state of the peer and drops down into the applicable loop
+func (rf *Raft) run() {
+	for {
+		switch {
+		case rf.killed:
+			DPrintf("Shutdown  🔫 ")
+			return
+		case rf.isFollower():
+			rf.runFollower()
+		case rf.isCandidate():
+			DPrintf("✋ Candidate state hit on %v", rf.me)
+			rf.runCandidate()
+		case rf.isLeader():
+			DPrintf("💪 leader state hit on %v", rf.me)
+			rf.runLeader()
+		}
+	}
+}
+
+func (rf *Raft) runFollower() {
+	heartbeatTimeout := randomTimeout(rf.heartbeatTimeout)
+
+	for {
+		DPrintf("👍🏻Follower state it on %v", rf.me)
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case <-heartbeatTimeout:
+			heartbeatTimeout = randomTimeout(rf.heartbeatTimeout)
+			DPrintf("⏲ follower %v election timed out", rf.me)
+			lastContact := rf.LastContact()
+			if time.Now().Sub(lastContact) < rf.heartbeatTimeout {
+				continue
+			}
+			rf.makeCandidate()
+			return
+		case <-rf.heartBeatChan:
+			DPrintf("💖 follower %v received heartbeat", rf.me)
+			continue
+		}
+	}
+}
+
+func (rf *Raft) runCandidate() {
+	electionTimeout := randomTimeout(rf.electionTimeout)
+	rf.incTerm()
+	votes := rf.dispatchVotes()
+
+	tally := 0
+	quorum := rf.getMajority()
+
+	for {
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case <-rf.heartBeatChan:
+			DPrintf("💖 Candidate %v received heartbeat", rf.me)
+			rf.makeFollower()
+			return
+		case <-electionTimeout:
+			DPrintf("⏲ Candidate %v election timed out", rf.me)
+			return
+		case vote := <-votes:
+			if vote.Term > rf.getTerm() {
+				rf.makeFollower()
+				rf.setTerm(vote.Term)
+			}
+			if vote.VoteGranted {
+				tally++
+			}
+			if tally >= quorum {
+				DPrintf("👺 I'm the boss %v", rf.me)
+
+				rf.makeLeader()
+				return
+			}
+		}
+	}
+}
```

Comment:
Candidates send out requestVote RPCs and then tally the resulting votes from other peers. They change their state to leader if the election succeeds.